proxy/src/main/java/com/wavefront/agent/listeners/ChannelStringHandler.java

```diff
@@ -39,17 +48,61 @@
   private final ReportableEntityPreprocessor preprocessor;
   private final PointHandler pointHandler;
 
+  private boolean logRawData = false;
+  /**
+   * Value of system property wavefront.proxy.lograwdata (for backwards compatibility)
+   */
+  private final boolean logRawDataFlag;
+  private double logRawDataRate;
+  private volatile long logRawUpdatedMillis = 0L;
+
   public ChannelStringHandler(Decoder<String> decoder,
                               final PointHandler pointhandler,
                               @Nullable final ReportableEntityPreprocessor preprocessor) {
     this.decoder = decoder;
     this.pointHandler = pointhandler;
     this.preprocessor = preprocessor;
+
+    // check the property setting for logging raw data
+    String logRawDataProperty = System.getProperty("wavefront.proxy.lograwdata");
+    logRawDataFlag = logRawDataProperty != null && logRawDataProperty.equalsIgnoreCase("true");
+    String logRawDataSampleRateProperty = System.getProperty("wavefront.proxy.lograwdata.sample-rate");
+    this.logRawDataRate = logRawDataSampleRateProperty != null &&
+        NumberUtils.isNumber(logRawDataSampleRateProperty) ? Double.parseDouble(logRawDataSampleRateProperty) : 1.0d;
+
+    // make sure the rate fits between 0.0d - 1.0d
+    if (logRawDataRate < 0.0d) {
+      logRawDataRate = 0.0d;
+    } else if (logRawDataRate > 1.0d) {
+      logRawDataRate = 1.0d;
+    }
   }
 
   @Override
   protected void channelRead0(ChannelHandlerContext ctx, String msg) throws Exception {
-    // if msg does not match metadata keywords then treat it as a point
+    // use data rate to determine sampling rate
+    // logging includes the source host and port
+    if (logRawUpdatedMillis + TimeUnit.SECONDS.toMillis(1) < System.currentTimeMillis()) {
+      if (logRawData != rawDataLogger.isLoggable(Level.FINEST)) {
+        logRawData = !logRawData;
+        rawDataLogger.info("Raw data logging is now " + (logRawData ?
+            "enabled with " + (logRawDataRate * 100) + "% sampling" :
+            "disabled"));
+      }
+      logRawUpdatedMillis = System.currentTimeMillis();
+    }
+
+    if ((logRawData || logRawDataFlag) &&
+        (logRawDataRate >= 1.0d || (logRawDataRate > 0.0d && RANDOM.nextDouble() < logRawDataRate))) {
+      if (ctx.channel().remoteAddress() != null) {
+        String hostAddress = ((InetSocketAddress) ctx.channel().remoteAddress()).getAddress().getHostAddress();
+        int localPort = ((InetSocketAddress) ctx.channel().localAddress()).getPort();
+        rawDataLogger.info(String.format("[%s>%d]%s", hostAddress, localPort, msg));
```

Comment:
Nit-pick: string concatenation is about 10x faster than String.format() so it might make sense to use concatenation here

Reply:
good idea!